0015-estimator-interface.md

```diff
@@ -0,0 +1,502 @@
+
+Moreover, the interface changes in this proposal will enable specific primitive implementations to enhance performance through reduced bandwidth on the runtime, compatibility with fast parameter binding, and multiplexing qubit-wise commuting observables.
+
+## Design Proposal <a name="design-proposal"></a>
+
+We will lead the design proposal with an example:
+
+```python
+# construct two circuits, the first with 2039 Parameters, the second with no Parameters
+circuit1 = QuantumCircuit(9)
+...
+circuit2 = QuantumCircuit(2)
+...
+
+# specify 128 different parameter value sets for circuit1, in a 4x32 shape
+parameter_values1 = np.random.uniform(size=(4, 32, 2039))
+# specify 4 observables to measure for circuit 1
+observables1 = ["ZZIIIIIII", "IZXIIIIII", "IIIIIIYZI", "IZXIIYIII"]
+
+# specify 18 observables to measure for circuit2
+observables2 = [Pauli("XYZIIII"), ..., SparsePauliOp({"YYIIIII": 0.2})]
+
+# invoke an estimator
+estimator = Estimator()
+job = estimator.run([
+  (circuit1, observables1, parameter_values1),
+  (circuit2, observables2)
+])
+
+# wait for the result
+job.result()
+
+# get a bundle of results for every task
+>> EstimatorResult(
+>>     TaskResult<{evs: ndarray<4, 32>, stds: ndarray<4, 32>}, local_metadata>, 
+>>     TaskResult<{evs: ndarray<18>, stds: ndarray<18>}, local_metadata>, 
+>>     global_metadata
+>> )
+```
+
+The above example shows various features of the design, including combining large numbers of observables with large numbers of parameter value sets for a single circuit.
+
+The detailed targetted signature, following the [deprecation period](#migration-path),
+is as follows:
+
+```python
+T = TypeVar("T", bound=Job)
+
+BindingsArrayLike = Union[
+    BindingsArray, 
+    NDArray, 
+    Dict[Parameter | Tuple[Parameter], NDArray]
+]
+
+ObservablesArrayLike = Union[
+    ObservablesArray,
+    Iterable[str | Pauli | SparsePauliOp]
+]
+
+EstimatorTaskLike = Union[
+    EstimatorTask,
+    Tuple[QuantumCircuit, ObservablesArrayLike, BindingsArrayLike]
+]
+
+class EstimatorBase(ABC, Generic[T]):
+    ...
+    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike], **options) -> T:
+        pass
+```
+
+### Migration Examples <a name="migration-examples"></a>
+
+Technical details of migration are discussed in the [Migration Path](#migration-path) section, but it is perhaps worthwhile to show some examples that demonstrate that, despite the new array and broadcasting abilities, the interface is not overly complicated for simple examples.
+
+One circuit, many observables:
+
+```python
+    # old API
+    estimator.run([circuit] * 4, observables=["ZIII", "IZII", "IIZI", "IIIZ"]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3, 0.4]), metadata={...})
+
+    # new API (we can use a single task)
+    estimator.run((circuit, ["ZIII", "IZII", "IIZI", "IIIZ"])).result()
+    >> [ResultTask({"evs": np.array([0.1, 0.2, 0.3, 0.4]), "stds": <...>}, metadata={...})]
```

Comment:
Agreed, we need better names here. The dataclass thing is in the works, and will make an appearance in the next RFC that @blakejohnson mentioned below, just haven't had a chance to update this PR yet.

Reply:
It is good to have more data fields as standard since metadata is arbitrary - in the existing Estimator it was not agreed on what stds was and was left at whatever was done, if anything, to something placed in metadata. That meant code could not easily be written to use such data since you would have to know about the specific implementation and what that provided in the metadata.